Support Retronix Sparrow Hawk board based on Renesas R-Car V4H SoC

DuyDang007 opened the pull request:
Hi everyone,

I would like to submit the code to support Retronix Sparrow Hawk board.

The board information can be found here:

https://www.renesas.com/en/support/partners/r-car-consortium/r-car-consortium-proactive-partner-list/rtx-sparrow-hawk?srsltid=AfmBOorxqVwVgVuRQ4Mi1N5FJYzQpdxNyP39oBwpRTHV3wtzNFtAVfCz

https://rcar-community.github.io/Sparrow-Hawk/index.html#introduction

Currently it supports only UART driver and confirmed with Hello World sample app. More features will be supported in the future

DuyDang007 committed:
board: rcar: Support Retronix Sparrow Hawk board

Add support for Retronix Sparrow Hawk board.
The board is based on R-Car V4H v3.0 (R8A779G3) SoC.

Signed-off-by: Duy Dang <duy.dang.yw@renesas.com>

DuyDang007 committed:
driver: clock_control: Add V4H clock control driver

Support Clock Pulse Generator driver for V4H

Signed-off-by: Duy Dang <duy.dang.yw@renesas.com>

DuyDang007 committed:
driver: pinctrl: Add support R-Car V4H SoC

Add pin definition for R-Car V4H SoC.

Signed-off-by: Duy Dang <duy.dang.yw@renesas.com>

DuyDang007 committed:
board: rcar: Add document for Retronix Sparrow Hawk board

Adds a page for Retronix Sparrow Hawk in Supported boards.

Signed-off-by: Duy Dang <duy.dang.yw@renesas.com>

nordicjm commented on dts/arm/renesas/rcar/gen4/r8a779g0.dtsi:
```diff
@@ -0,0 +1,60 @@
+			interrupts = <GIC_SPI 246 IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY>;
+			clocks = <&cpg CPG_MOD 514>, <&cpg CPG_CORE R8A779G0_CLK_SASYNCPERD1>;
+		};
+		hscif1: serial@e6550000 {
```
missing newline between nodes, fix in whole PR

DuyDang007 commented on soc/renesas/rcar/rcar_gen4/Kconfig.defconfig.r8a779g0:
```diff
@@ -0,0 +1,16 @@
+	default "r8a779g0" if SOC_R8A779G0_R52
+
+config ARCH
+	default "arm"
```
I removed this

DuyDang007 commented on boards/retronix/sparrowhawk_rcar_v4h/sparrowhawk_rcar_v4h_r8a779g0_r52_defconfig:
```diff
@@ -0,0 +1,19 @@
+CONFIG_CLOCK_CONTROL=y
+# Generic timer clock frequency in Hz
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=16666667
```
I moved it to `Kconfig.defconfig.r8a779g0`.

Could you explain "and must come from dts property using functions"? I see that the driver read from this config directly, not the dts

kartben commented on boards/retronix/sparrowhawk_rcar_v4h/doc/img/01_v4h_sparrow_hawk.jpg:
please use the following webp / transparent bg / better quality image instead, name the same as board too so "sparrowhawk_rcar_v4h.webp"

![Image](https://github.com/user-attachments/assets/b8c53ef4-fd69-4c21-9c88-9f3410a8208f)

kartben commented on boards/retronix/index.rst:
```diff
@@ -0,0 +1,10 @@
+.. _boards-retronix:
+
+Retronix Technology Inc
```
you need to add a new entry to dts/bindings/vendor-prefixes.txt

kartben replied:
yes - sorry! and thanks :)

kartben commented on boards/retronix/sparrowhawk_rcar_v4h/doc/sparrow_hawk_rcar_v4h_r52.rst:
```diff
@@ -0,0 +1,159 @@
+
+.. figure:: img/02_v4h_block_diagram.jpg
+   :align: center
+   :alt: R-Car V4H block diagram
```
please consider just linking to the product webpage instead, I don't think this _needs_ to be in Zephyr documentation. In any case, jpg shouldn't be used for diagram - prefer webp.

kartben commented on boards/retronix/sparrowhawk_rcar_v4h/doc/sparrow_hawk_rcar_v4h_r52.rst:
```diff
@@ -0,0 +1,159 @@
+
+.. zephyr:board-supported-hw::
+
+More features will be supported soon.
```
```suggestion
```

DuyDang007 replied:
I removed it.

kartben commented on boards/retronix/sparrowhawk_rcar_v4h/board.yml:
```diff
@@ -0,0 +1,6 @@
+board:
+  name: sparrowhawk_rcar_v4h
+  full_name: Retronix Sparrow Hawk
```
```suggestion
  full_name: Sparrow Hawk R-Car V4H SBC
```

DuyDang007 replied:
I updated the name

kartben reviewed (dismissed):
+1 for docs, thanks!

DuyDang007 commented on boards/retronix/sparrowhawk_rcar_v4h/doc/sparrow_hawk_rcar_v4h_r52.rst:
```diff
@@ -0,0 +1,148 @@
+   U-Boot 2025.07 (Aug 07 2025 - 04:02:12 +0000)
+
+   CPU:   Renesas Electronics R8A779G0 rev 3.0
+   Model: Retronix Sparrow Hawk board based on r8a779g3
```
This is the boot log, we keep it as is.

There is some SoC variant R8A779Gx, they have minor differences that does not affect to our software. So I keep it as R8A779G0

nordicjm commented on dts/arm/renesas/rcar/gen4/r8a779g0.dtsi:
```diff
@@ -0,0 +1,61 @@
+		};
+
+		/* Using domain 0
+		 */
```
why does this comment need to space over 2 lines?

nordicjm commented on dts/arm/renesas/rcar/gen4/r8a779g0.dtsi:
```diff
@@ -0,0 +1,61 @@
+		pfc: pin-controller@e6050000 {
+			compatible = "renesas,rcar-pfc";
+			reg = <0xe6050000 0x1d8>, <0xe6050800 0x1d8>,
+				  <0xe6058000 0x1d8>, <0xe6058800 0x1d8>,
```
bad alignment

nordicjm commented on boards/retronix/sparrowhawk_rcar_v4h/sparrowhawk_rcar_v4h_r8a779g0_r52_defconfig:
```diff
@@ -0,0 +1,14 @@
+CONFIG_UART_CONSOLE=y
+
+# GPIO
+CONFIG_GPIO=n
```
should not be needed

kylebonnici commented on dts/arm/renesas/rcar/gen4/r8a779g0.dtsi:
Diff to fix formatting
```diff
@@ -12,20 +12,20 @@
 	soc {
 		pfc: pin-controller@e6050000 {
 			compatible = "renesas,rcar-pfc";
 			reg = <0xe6050000 0x1d8>, <0xe6050800 0x1d8>,
-				  <0xe6058000 0x1d8>, <0xe6058800 0x1d8>,
-				  <0xe6060000 0x1d8>, <0xe6060800 0x1d8>,
-				  <0xe6061000 0x1d8>, <0xe6061800 0x1d8>,
-				  <0xe6068000 0x1d8>, <0xe6078000 0x0e8>;
+			      <0xe6058000 0x1d8>, <0xe6058800 0x1d8>,
+			      <0xe6060000 0x1d8>, <0xe6060800 0x1d8>,
+			      <0xe6061000 0x1d8>, <0xe6061800 0x1d8>,
+			      <0xe6068000 0x1d8>, <0xe6078000 0x0e8>;
 		};
 
 		/* Using domain 0
 		 */
 		cpg: clock-controller@e6150000 {
 			compatible = "renesas,r8a779g0-cpg-mssr";
 			reg = <0xe6150000 0x8a4>,  /* GPG */
-				  <0xe6150000 0x2e74>; /* Module standby, software reset */
+			      <0xe6150000 0x2e74>; /* Module standby, software reset */
 			#clock-cells = <2>;
 		};
 
 		gpio0: gpio@e6050000 {

```

DuyDang007 commented on soc/renesas/rcar/rcar_gen4/Kconfig.defconfig.r8a779g0:
```diff
@@ -0,0 +1,18 @@
+
+# Arm Generic Timer frequency (Hz)
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 16666667
```
This config is used in arm_arch_timer.c for system tick. The driver read CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC value directly so I cannot make use of DT function.

In V4H SoC, 16.667MHz is the fixed frequency of Generic Timer

DuyDang007 replied:
So, I will add the frequency value to its timer node like below:

rcar_gen4_v4h_cr52.dtsi:
```
	timer {
		compatible = "arm,armv8-timer";
		interrupt-parent = <&gic>;
		interrupts = <GIC_PPI 13 IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY>,
		             <GIC_PPI 14 IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY>,
		             <GIC_PPI 11 IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY>,
		             <GIC_PPI 10 IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY>;
+		clock-frequency = <0xfe502b>;
	};
```

Is it correct?

kartben commented:
re-assigning to renesas r-car maintainer

DuyDang007 commented on dts/arm/renesas/rcar/gen4/rcar_gen4_v4h_cr52.dtsi:
```diff
@@ -0,0 +1,68 @@
+		compatible = "arm,armv8-timer";
+		interrupt-parent = <&gic>;
+		interrupts = <GIC_PPI 13 IRQ_TYPE_LEVEL
+					IRQ_DEFAULT_PRIORITY>,
```
I fixed the alignment issues.

DuyDang007 commented on boards/retronix/sparrowhawk_rcar_v4h/sparrowhawk_rcar_v4h_r8a779g0_r52.dts:
```diff
@@ -0,0 +1,42 @@
+};
+
+&gpio0 {
+	status = "disabled";
```
Removed. It's a duplicated "disabled"

DuyDang007 commented on dts/arm/renesas/rcar/gen4/rcar_gen4_v4h_cr52.dtsi:
This causes the error with compliance check due to mixed tab and space indent:

```
 CODE_INDENT: code indent should use tabs where possible
File:dts/arm/renesas/rcar/gen4/rcar_gen4_v4h_cr52.dtsi
Line:28
 CODE_INDENT: code indent should use tabs where possible
File:dts/arm/renesas/rcar/gen4/rcar_gen4_v4h_cr52.dtsi
Line:29
 CODE_INDENT: code indent should use tabs where possible
File:dts/arm/renesas/rcar/gen4/rcar_gen4_v4h_cr52.dtsi
Line:30
```

DuyDang007 committed:
soc: renesas: rcar: Support V4H SoC

Add initial support for Renesas R-Car V4H SoC.
Support Zephyr RTOS on Cortex R52, 1.4GHz core.
For more information and documentation, please visit the product page:
https://www.renesas.com/en/products/r-car-v4h

Signed-off-by: Duy Dang <duy.dang.yw@renesas.com>

khoa-nguyen-18 reviewed (commented):
Hi, this is my first time looking at the RCAR. If any of my information is incorrect, please feel free to correct me :)

khoa-nguyen-18 reviewed (dismissed):
The SoC and board support look good to me

nhutnguyenkc commented on dts/bindings/timer/arm,armv8-timer.yaml:
```diff
@@ -7,3 +7,5 @@ include: base.yaml
   interrupts:
     required: true
+  clock-frequency:
+    type: int
```
It seems incorrect. I saw that you add this property to assign a value to `SYS_CLOCK_HW_CYCLES_PER_SEC`, but it does not work in that way. You should define a node in dtsi for the clock source supplying to generic timer and get its value from it. I check the user manual and believe that it should be CL16M clock.

Although the clock control is supported, I cannot find any related clock node in dtsi, at least a node for external clock source should be provided. Please revert this change.

```suggestion
```

DuyDang007 replied:
Listing all the clock sources for clock controller driver is good idea. However, I want to keep the current implementation to align with other R-Car devices. It might be improved later.

I will add one node for CL16_RT clock to avoid adding clock-frequency in bindings.

DuyDang007 replied:
Hi @nhutnguyenkc , I added a node for clock to avoid modifying binding file. Just for Generic timer only.

Thanks for your advice

nhutnguyenkc replied:
I confirmed it

nhutnguyenkc reviewed (changes requested):
Great to see other Renesas platforms get support.

DuyDang007 commented:
Hello @lorc , @khoa-nguyen-18 , @nhutnguyenkc ,

Sorry if I don't understand the working flow. Is there any item need to be done to make this merged?

drozd-svtln commented on drivers/clock_control/clock_control_r8a779g0_cpg_mssr.c:
```diff
@@ -0,0 +1,282 @@
+		ret = rcar_cpg_mstp_clock_endisable(DEVICE_MMIO_GET(dev), clk->module, enable);
+		k_spin_unlock(&data->cmn.lock, key);
+	} else if (clk->domain == CPG_CORE) {
+		ret = r8a779g0_cpg_core_clock_endisable(dev, clk, enable);
```
It is not obvious why you use k_spin for `rcar_cpg_mstp_clock_endisable` but not for `r8a779g0_cpg_core_clock_endisable`. 
It is better to remove `r8a779g0_cpg_core_clock_endisable` function and just move its functionality directly here. And after renaming of `r8a779g0_cpg_enable_disable_core` we will have here something like 
```
} else if (clk->domain == CPG_CORE) {
    // finding clock, setting rate etc from r8a779g0_cpg_core_clock_endisable
    key = k_spin_lock(&data->cmn.lock);
    // call of r8a779g0_cpg_core_clock_endisable
    k_spin_unlock(&data->cmn.lock, key);
} else {
```

drozd-svtln commented on drivers/clock_control/clock_control_r8a779g0_cpg_mssr.c:
```diff
@@ -0,0 +1,282 @@
+	return 0;
+}
+
+int r8a779g0_cpg_mssr_start_stop(const struct device *dev, clock_control_subsys_t sys, bool enable)
```
Rename `start_stop` to `on_off` to be consistent with other changes

khoa-nguyen-18 commented:
Hi @lorc, could you please help to review this PR?

khoa-nguyen-18 commented:
Hi @kartben, since the assignee of this PR seems to be busy at the moment, should we add another Renesas maintainer as an assignee for this PR?

klogg commented:
> Hi @kartben, since the assignee of this PR seems to be busy at the moment, should we add another Renesas maintainer as an assignee for this PR?

Hey @khoa-nguyen-18 please check comments from @drozd-svtln, they need to be addressed

DuyDang007 commented on boards/retronix/sparrowhawk_rcar_v4h/board.cmake:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
```
Updated

drozd-svtln commented:
I don't know why, but it is failing to unresolve the thread. Please pay attention to the one remaining comment:
https://github.com/zephyrproject-rtos/zephyr/pull/97783/changes#r2871747809

DuyDang007 commented:
Hi @drozd-svtln 

I created new PR to support Counter driver for V4H Sparrowhawk: https://github.com/zephyrproject-rtos/zephyr/pull/106796

I cannot mention you in that PR. So, I write it here. Could you please support to review it?
